### PR TITLE
Buyandsellstocks

### LINKED ISSUE #4569: [FEATURE REQUEST] Buy And Sell Stocks Based DP problems Missing
### What would you like to Propose?

I explored the DP Section ( thealgorithms/dynamicprogramming/ ) and wasn't  able to find Buy and Sell Stocks pattern based DP problems. Would be better if you we could add those problems to the repo
 
The problem involves buying and selling stocks with some constraints and achieving the maximum profit from it
Do Check out the problems

I would like to Contribute to it as my first hacktoberfest Contribution. So can you pls assign it to me , I have The code ready.

### Issue details

To know more about this you can search for leetcode Buy and sell stock problems

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by BharathSubu**
Adding Buy and Sell Stocks
Fixes #4569

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.

### CODE REVIEW

File: src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+// https://en.wikipedia.org/wiki/Dynamic_programming
+
+/*
+ * You are given an integer array prices where prices[i] is the price of a given
+ * stock on the ith day.
+ * On each day, you may decide to buy and/or sell the stock. You can only hold
+ * at most one share of the stock at any time. However, you can buy it then
+ * immediately sell it on the same day.
+ * Find and return the maximum profit you can achieve.
```
Comment (Manan-09) [inline src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java:18]: @BharathSubu  can we buy and sell stock as many times as we want?
If that is the case , we can find maximum profit in O(n) TC and O(1) SC , using greedy approch

File: src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        int prices[] = new int[n];
+        for (int i = 0; i < n; i++) {
+            prices[i] = sc.nextInt();
+        }
+        // memoization
+        System.out.println(buyAndSellStockMemo(prices));
+        // Tabulation
+        System.out.println(buyAndSellStockMemo(prices));
+        sc.close();
+    }
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java:34]: This is not needed.
```suggestion
```

File: src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
+package com.thealgorithms.dynamicprogramming;
+
+import java.util.Arrays;
+import java.util.Scanner;
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java:4]:
```suggestion
```

File: src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * @author Bharath Subu (https://github.com/BharathSubu)
+ */
+
+// https://en.wikipedia.org/wiki/Dynamic_programming
```
Comment (vil02) [inline src/main/java/com/thealgorithms/dynamicprogramming/BuyAndSellStock.java:10]: Please add an exact link with the description of the problem and its solution.

**Review by vil02 — CHANGES_REQUESTED** [review]
@BharathSubu please add tests for the provided implementation.

Some examples to look at:
- [`FactorialTest.java`](https://github.com/TheAlgorithms/Java/blob/master/src/test/java/com/thealgorithms/maths/FactorialTest.java),
- [`SecondMinMaxTest.java`](https://github.com/TheAlgorithms/Java/blob/master/src/test/java/com/thealgorithms/maths/SecondMinMaxTest.java).

**Pull request closed without merging**